sig-api-machinery/extensible-admission-to-beta.md

```diff
@@ -0,0 +1,292 @@
+it less error-prone, easier to configure, and more flexible for admission plugin
+developers.
+
+## Requirements for extensible admission
+Ordering the admission chain is critical.  One mutating admission plugin can invalidate
+the results of other admission plugins.  For instance, the `LimitVerifier` would
+invalidate results for the `ResourceQuota` plugin.  More generally, object mutations
+must happen before object validations.  In addition, compiled validating admission plugins
+must run after mutating admission webhooks.  Splitting the chain into mutators and validators
+will resolve the coarse grained ordering problem and would require two different webhook APIs
+since the `Status` would be different for both now and we can reasonably expect different
+evolutions.  It would also force changes in the registration APIs since ordering matters
+for individual mutators, but ordering does not matter for validators.  Having fewer items to 
+order makes the registration API easier to reason about for maintainability.
+
+Initializers as implemented today attempt to push new API validation requirements into the 
+strategy and new requirements onto the admission plugins handling updates and creates.  The
+former is unnecessary and the latter doesn't work.
+
+API validation on uninitialized objects can reasonably follow the rules for validation on create
+plus `ValidateObjectMetaUpdate` (to prevent name and UID changes and the like).
+Since the object hasn't been "ready", there is no reason to add the validation rules required on
+updates since restrictions like field immutability don't logically apply and would restrict the power
+of admission extensions to the point of unusability.  Using the results for creation validation on 
+uninitialized objects and objects transitioning from uninitialized to initialized lets the
+strategy author focus on their API needs.
+
+Admission validation is also needed on resources being initialized.  During initialization, there
+may be some admission validation required, but it isn't normal API update admission.  The update
+admission mutators shouldn't run and since the subject (user.Info) on the admission attributes is
+NOT the original user, normal update admission mutators and validators are likely to make improper
+choices.  When the object transitions from uninitialized to initialized state, the original object
+creator should be artificially injected into the admission attributes and the create admission 
+validators should run.  This allows mechanisms like PSP to function properly during their validation
+phase.  Such injection requires persisting the original user.Info (probably encrypted) into the 
+object.
+
+## How much is needed for beta
+Basically all of it.  After considering the needs of initializers, the admission chain needs to
+be split into mutators and validators.  That has a direct impact on the webhook admission registration
```

Comment:
Would mutators also be allowed to validate, or would it be completely separate? I'm an advocate for breaking out the defaulting (mutation) from the PSP, but as it currently functions it's incompatible with this split.

Reply:
> Would mutators also be allowed to validate, or would it be completely separate?

If it isn't completely separate, then such a plugin cannot observe the final object.  Other mutators in the direct chain and all initializers will happen after it has validated.